src/staking/block_reward_validator.cpp

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2019 The Unit-e developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://opensource.org/licenses/MIT.
+
+#include <staking/block_reward_validator.h>
+
+#include <chain.h>
+#include <consensus/validation.h>
+#include <primitives/transaction.h>
+
+#include <utilmoneystr.h>
+
+namespace staking {
+
+class BlockRewardValidatorImpl : public BlockRewardValidator {
+ private:
+  Dependency<blockchain::Behavior> m_behavior;
+
+ public:
+  BlockRewardValidatorImpl(
+      Dependency<blockchain::Behavior> behavior)
+      : m_behavior(behavior) {}
+
+  bool CheckBlockRewards(const CTransaction &coinbase_tx, CValidationState &state, const CBlockIndex &index,
+                         CAmount input_amount, CAmount fees) override {
+    assert(MoneyRange(input_amount));
+    assert(MoneyRange(fees));
+
+    CAmount total_reward = fees + m_behavior->CalculateBlockReward(index.nHeight);
+
+    std::size_t num_reward_outputs = 1;
+
+    CAmount output_amount = coinbase_tx.GetValueOut();
```

Comment:
I think checking if reward outputs exist (`bad-cb-no-reward`) should be recreated.
Also we can check if the coinbase has at least one extra output (the stake).
https://github.com/dtr-org/unit-e/issues/777

Reply:
> Also we can check if the coinbase has at least one extra output (the stake).

I didn't do it purposefully so proposers would be able to combine stake and reward to avoid increasing the number of UTXOs.

> I think checking if reward outputs exist (bad-cb-no-reward) should be recreated.

I can add check `coinbase_tx.vout.size() < num_reward_outputs` from
https://github.com/dtr-org/unit-e/pull/1081/files#diff-6af7edfed7e43b88c98882bf45450cd1R36

Reply:
> I can add check coinbase_tx.vout.size() < num_reward_outputs from
https://github.com/dtr-org/unit-e/pull/1081/files#diff-6af7edfed7e43b88c98882bf45450cd1R36

Sounds like a great solution.

Reply:
Done.